// @stadtkatalog/openinghours/lib/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Possible formats for the representation of a weekday.
 * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/RelativeTimeFormat/RelativeTimeFormat#Syntax
 */
var WeekdayFormat;
(function (WeekdayFormat) {
    /**
     * `long` e.g. "Thursday"
     */
    WeekdayFormat["long"] = "long";
    /**
     * `short` e.g. "Thu"
     */
    WeekdayFormat["short"] = "short";
    /**
     * `narrow` e.g. "T"
     */
    WeekdayFormat["narrow"] = "narrow";
})(WeekdayFormat = exports.WeekdayFormat || (exports.WeekdayFormat = {}));
/**
 * The type of a range.
 */
var RangeType;
(function (RangeType) {
    /**
     * Weekday time ranges are from Monday to Sunday.
     */
    RangeType["weekday"] = "weekday";
    /**
     * Time range for Holidays.
     */
    RangeType["holiday"] = "holiday";
    /**
     * Time ranges for specific special days.
     */
    RangeType["special"] = "special";
})(RangeType = exports.RangeType || (exports.RangeType = {}));
