// @lets-talk/dialogflow-fulfillment/src/constants.js
exports.contextNameBotHasTransfered = 'DERIVADO_BOT';

exports.contextBotHasTransfered = {
    'name': this.contextNameBotHasTransfered,
    'lifespan': 10,
    'parameters': {}
};

exports.contextNewTopic = {
    'name': 'NUEVO_TEMA',
    'lifespan': 2,
    'parameters': {}
};

exports.serviceName = "dialogflow-fulfillment";

exports.requiredSettings = [
    "token",
    "organization",
    "project",
    "environment",
    "availability",
    "adaptorex_token",
    "df_client_email",
    "df_private_key",
]

exports.providers = {
    twilio: "twilio",
    lt: "letstalk_agent",
}

exports.typing = {
    content: "",
    mtype: "SYSTEM",
    internal: false,
    content_type: "typing",
  };


exports.mockedResponse = '<?xml version="1.0" encoding="utf-8"?><soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema"><soap:Body><ConsultaResponse xmlns="http://sinacofi.cl/WebServices"><ConsultaResult><codigoRetorno>10000</codigoRetorno><RESUMEN><RUT xmlns="http://wsdl.sinacofi.cl/SDN121RES">153742642</RUT><USUARIO xmlns="http://wsdl.sinacofi.cl/SDN121RES">LETICP  </USUARIO><IDCHALLENGE xmlns="http://wsdl.sinacofi.cl/SDN121RES">202012061400070966499828</IDCHALLENGE></RESUMEN><DETALLE><DESAFIO xmlns="http://wsdl.sinacofi.cl/SDN121RES"><PREGUNTAS><CODIGO_PREGUNTA>1         </CODIGO_PREGUNTA><PREGUNTA>¿A Qué Sucursal de la Región Metropolitana fue este mes de Mayo?</PREGUNTA><ALTERNATIVAS><CODIGO_RESPUESTA>1         </CODIGO_RESPUESTA><RESPUESTA>Sucursal CUPRUM Conchali, Santiago</RESPUESTA></ALTERNATIVAS><ALTERNATIVAS><CODIGO_RESPUESTA>2         </CODIGO_RESPUESTA><RESPUESTA>Sucursal CUPRUM Est. central, Santiago</RESPUESTA></ALTERNATIVAS><ALTERNATIVAS><CODIGO_RESPUESTA>3         </CODIGO_RESPUESTA><RESPUESTA>Sucursal CUPRUM Lo prado, Santiago</RESPUESTA></ALTERNATIVAS><ALTERNATIVAS><CODIGO_RESPUESTA>4         </CODIGO_RESPUESTA><RESPUESTA>Sucursal CUPRUM La cisterna, Santiago</RESPUESTA></ALTERNATIVAS></PREGUNTAS><PREGUNTAS><CODIGO_PREGUNTA>REQ0002   </CODIGO_PREGUNTA><PREGUNTA>¿Está o estuvo su domicilio en alguna de las siguientes direcciones?</PREGUNTA><ALTERNATIVAS><CODIGO_RESPUESTA>RSP0688   </CODIGO_RESPUESTA><RESPUESTA>M RODRIGUEZ 39  CHIGUAYANTE</RESPUESTA></ALTERNATIVAS><ALTERNATIVAS><CODIGO_RESPUESTA>RSP2724   </CODIGO_RESPUESTA><RESPUESTA>ALMIRANTE GOMEZ CARRE¬O19</RESPUESTA></ALTERNATIVAS><ALTERNATIVAS><CODIGO_RESPUESTA>RSP4840   </CODIGO_RESPUESTA><RESPUESTA>CORR STA JUANA</RESPUESTA></ALTERNATIVAS><ALTERNATIVAS><CODIGO_RESPUESTA>RSP4886   </CODIGO_RESPUESTA><RESPUESTA>PARAGUAY 2091</RESPUESTA></ALTERNATIVAS></PREGUNTAS><PREGUNTAS><CODIGO_PREGUNTA>REQ0008   </CODIGO_PREGUNTA><PREGUNTA>¿Cuál es el apellido materno de su padre?</PREGUNTA><ALTERNATIVAS><CODIGO_RESPUESTA>RSP1547   </CODIGO_RESPUESTA><RESPUESTA>PARRA</RESPUESTA></ALTERNATIVAS><ALTERNATIVAS><CODIGO_RESPUESTA>RSP5489   </CODIGO_RESPUESTA><RESPUESTA>NINGUNA DE LAS ANTERIORES</RESPUESTA></ALTERNATIVAS><ALTERNATIVAS><CODIGO_RESPUESTA>RSP9030   </CODIGO_RESPUESTA><RESPUESTA>TORO</RESPUESTA></ALTERNATIVAS><ALTERNATIVAS><CODIGO_RESPUESTA>RSP9526   </CODIGO_RESPUESTA><RESPUESTA>HERNÁNDEZ</RESPUESTA></ALTERNATIVAS></PREGUNTAS><PREGUNTAS><CODIGO_PREGUNTA>REQ0023   </CODIGO_PREGUNTA><PREGUNTA>¿Está o estuvo alguna vez su dirección en una de las siguientes comunas?</PREGUNTA><ALTERNATIVAS><CODIGO_RESPUESTA>RSP0963   </CODIGO_RESPUESTA><RESPUESTA>ISLA DE MAIPO</RESPUESTA></ALTERNATIVAS><ALTERNATIVAS><CODIGO_RESPUESTA>RSP4115   </CODIGO_RESPUESTA><RESPUESTA>LA REINA</RESPUESTA></ALTERNATIVAS><ALTERNATIVAS><CODIGO_RESPUESTA>RSP8243   </CODIGO_RESPUESTA><RESPUESTA>SAN RAMON</RESPUESTA></ALTERNATIVAS><ALTERNATIVAS><CODIGO_RESPUESTA>RSP9317   </CODIGO_RESPUESTA><RESPUESTA>LA FLORIDA</RESPUESTA></ALTERNATIVAS></PREGUNTAS></DESAFIO></DETALLE></ConsultaResult></ConsultaResponse></soap:Body></soap:Envelope>'
exports.SinacofiSelectedAnswer = "selected_answer";

exports.headers = {
    content_types: {
        xml: "text/xml",
    },
    soap_actions: {
        sinacofi: "http://sinacofi.cl/WebServices/Consulta",
    },
}

exports.errorMessages = {
    unknownProvider: "This provider is currently not supported.",
}
